### PR TITLE
feat: implement node runtime region awareness for cdk vended custom resources

### PR DESCRIPTION
**Pull request by colifran**
### Reason for this change

This PR introduces node runtime region awareness into the custom resource provider framework by doing the following:

- Adding a `LATEST_NODE_RUNTIME_MAP` fact table used to maintain the latest node runtime available per AWS partition
- Introducing `determineLatestNodeRuntime` and `determineLatestNodeRuntimeName` functions which determine the latest Lambda node `Runtime` and the latest Lambda node `Runtime` name, respectively
- Updating the custom resource provider framework to utilize these two functions when code generating the runtime property for providers using node runtimes

### Description of changes

This PR can be segmented into the following code changes:

- A `LATEST_NODE_RUNTIME_MAP` fact table was added which maintains the latest Lambda node runtime available per AWS partition.
- Introduced `determineLatestNodeRuntime` and `determineLatestNodeRuntimeName` functions which determine the latest Lambda node `Runtime` and the latest Lambda node `Runtime` name, respectively.
- The existing runtime property being code generated via the custom resource provider framework has been altered to now use the appropriate runtime determiner function, i.e., `determineLatestNodeRuntimeName` for `CustomResourceProvider` or `determineLatestNodeRuntime` for `Function` or `SingletonFunction`. Any custom resource provider using a python runtime will not use either of these functions.
- To consolidate and manage importing of external modules a `ModuleImporter` class has been created. This class allows external modules to be registered as an import for a target module and prevents duplicate imports for modules that contain multiple framework components. This class also provides the ability to specify different import paths which allows all external modules to be consolidated into a single class rather than having duplicate external modules defined for different import paths. Lastly, this class determines whether or not external modules should be imported selectively or if all targets in the external module should be imported under an alias, i.e., `import { Function } from 'aws-lambda'` vs. `import * as lambda from 'aws-lambda'`
- A `CallableExpr` class was created to allow expression proxies to be created from a specified expression name. This allows the new runtime determiner functions and other module specific functions to be called from their specified module and built into a JavaScript object that will mirror the JavaScript operations done to it in an expression tree.

### Description of how you validated changes

Manually tested all individual custom resource handlers for create, update, and delete. I verified that the behavior seen was what was expected based off of the handler code. All impacted integ tests were updated. New unit tests were added to the custom resource handler framework to test that generated code correctly included `determineLatestNodeRuntime` for Lambda based handlers and `determineLatestNodeRuntimeName` for `CfnResource` based handlers. Added new unit tests to test the functionality of `determineLatestNodeRuntime` and `determineLatestNodeRuntimeName` for region agnostic stacks, stacks in commercial region, stacks in China regions, stacks in ADC regions, and stacks in GovCloud regions.

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/custom-resource-handlers/lib/custom-resources-framework/classes.ts
Comment (colifran) [inline packages/@aws-cdk/custom-resource-handlers/lib/custom-resources-framework/classes.ts:220]: This is one example of what the `ModuleImporter` class helps with. We can maintain one module defined, i.e., `CORE_MODULE`, with all of it's types, expressions, etc. rather than needing to define a duplicate module based on where the module would be imported from. Instead, when registering an import we can just register the same module with a different `fromLocation` specified. Now the code looks a lot cleaner.

File: packages/@aws-cdk/custom-resource-handlers/lib/custom-resources-framework/classes.ts
Comment (colifran) [inline packages/@aws-cdk/custom-resource-handlers/lib/custom-resources-framework/classes.ts:321]: Lots of reduced code here due to `ModuleImporter` class managing imports

Comment (colifran) [inline packages/@aws-cdk/custom-resource-handlers/lib/custom-resources-framework/classes.ts:321]: Lots of reduced code here due to `ModuleImporter` class managing imports

File: packages/aws-cdk-lib/region-info/lib/fact.ts
Comment (rix0rrr) [inline packages/aws-cdk-lib/region-info/lib/fact.ts:201]: I also would expect this version to be used for `lambda.Runtime.LATEST_NODEJS`

File: packages/aws-cdk-lib/custom-resources/lib/provider-framework/provider.ts
Comment (rix0rrr) [inline packages/aws-cdk-lib/custom-resources/lib/provider-framework/provider.ts:11]: Explain to me why this code needed to be generated and couldn't just be a single handwritten function?

As far as I can tell, there are no parameters that go into the generation?

**Review by rix0rrr — COMMENTED** [review]
> The overall name of the framework was changed from HandlerFramework to ProviderFramework. This is because the framework is being used to code generate custom resource providers. The custom resource providers execute handler code that this framework has the responsibility of bundling, but calling this a HandlerFramework isn't accurate. 

Not in fact true, as far as I'm aware. We call it "custom resource handlers" etc, but I'm pretty sure it also contains Lambda Handlers that don't go into custom resources. Yes, 90% of them are custom resources, but not 100%. So calling this all "custom resources" is a mental shortcut.

File: packages/aws-cdk-lib/custom-resources/lib/provider-framework/provider.ts
Comment (colifran) [reply packages/aws-cdk-lib/custom-resources/lib/provider-framework/provider.ts:11]: Going to hand write the two functions `determineLatestNodeRuntime` and `determineLatestNodeRuntimeName` as discussed offline.

**Review by MrArnoldPalmer — COMMENTED** [review]
Lot's of good improvements to the codegen here that makes it more robust. I'll leave others to approve but I think this handles the move back to a "deploy time" computed runtime value well. 👍🏻

**Review by scanlonp — COMMENTED** [review]
Looking good, would like to go over it with you in more depth. Only small thing I noticed is we should add `isVariable: true` to the runtime created by the `determineLatestNodeRuntime()` function in `runtime.ts`.

File: packages/@aws-cdk/custom-resource-handlers/lib/custom-resources-framework/classes.ts
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/custom-resource-handlers/lib/custom-resources-framework/classes.ts:371]: I'm seeing paths hardcoded a lot. Can we turn most of these into a const?

Comment (TheRealAmazonKendra) [reply packages/@aws-cdk/custom-resource-handlers/lib/custom-resources-framework/classes.ts:371]: We could unintentionally break ourselves if we ever move packages around.

File: packages/@aws-cdk/custom-resource-handlers/lib/custom-resources-framework/modules.ts
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/custom-resource-handlers/lib/custom-resources-framework/modules.ts:62]: All the hardcoded paths continue to worry me.

File: packages/@aws-cdk/custom-resource-handlers/test/custom-resources-framework/expected/node-runtime/custom-resource-provider-core.ts
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/custom-resource-handlers/test/custom-resources-framework/expected/node-runtime/custom-resource-provider-core.ts:1]: NO! BAD! Obey prettier.

Comment (TheRealAmazonKendra) [reply packages/@aws-cdk/custom-resource-handlers/test/custom-resources-framework/expected/node-runtime/custom-resource-provider-core.ts:1]: Wait, are these generated?

File: packages/@aws-cdk/custom-resource-handlers/test/custom-resources-framework/expected/python-runtime/custom-resource-provider-core.ts
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/custom-resource-handlers/test/custom-resources-framework/expected/python-runtime/custom-resource-provider-core.ts:29]: node to python? Seems sketch.

Comment (TheRealAmazonKendra) [reply packages/@aws-cdk/custom-resource-handlers/test/custom-resources-framework/expected/python-runtime/custom-resource-provider-core.ts:29]: Oh wait, this is a test for python runtimes. How did we end up with node here? Ooooops.

File: packages/@aws-cdk/custom-resource-handlers/lib/custom-resources-framework/callable-expr.ts
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/custom-resource-handlers/lib/custom-resources-framework/callable-expr.ts:8]: This is a comment for almost all the files in this folder. I'm not seeing a lot of tests here. I'd like to see more unit tests added.

File: packages/aws-cdk-lib/aws-lambda/test/function.test.ts
Comment (TheRealAmazonKendra) [inline packages/aws-cdk-lib/aws-lambda/test/function.test.ts:4155]: A test case for ADC is missing

Comment (TheRealAmazonKendra) [reply packages/aws-cdk-lib/aws-lambda/test/function.test.ts:4155]: Actually below I see that this is actually for ADC. GovCloud is the one that is actually missing here.

File: packages/aws-cdk-lib/core/lib/custom-resource-provider/custom-resource-provider.ts
Comment (TheRealAmazonKendra) [inline packages/aws-cdk-lib/core/lib/custom-resource-provider/custom-resource-provider.ts:141]: Can we either add deprecation to the older runtimes or remove them completely as you can no longer create or update functions that contain them (at least for 12)? https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html#runtimes-deprecated

File: packages/aws-cdk-lib/core/test/custom-resource-provider/custom-resource-provider.test.ts
Comment (TheRealAmazonKendra) [inline packages/aws-cdk-lib/core/test/custom-resource-provider/custom-resource-provider.test.ts:627]: ADC !== GovCloud

File: packages/aws-cdk-lib/region-info/build-tools/fact-tables.ts
Comment (TheRealAmazonKendra) [inline packages/aws-cdk-lib/region-info/build-tools/fact-tables.ts:3458]: Default cannot be node20, it must be node18

File: packages/aws-cdk-lib/aws-lambda/test/function.test.ts
Comment (TheRealAmazonKendra) [reply packages/aws-cdk-lib/aws-lambda/test/function.test.ts:4155]: I would also like to see a test for a region that we don't currently have in the map to show that we are properly defaulting to node18 on those.

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
On the first pass through this is looking pretty good. Please see my comments inline. Also a had a whole bunch of questions inline as well.

File: packages/aws-cdk-lib/region-info/build-tools/fact-tables.ts
Comment (colifran) [reply packages/aws-cdk-lib/region-info/build-tools/fact-tables.ts:3458]: This is not named well but it is not the default. `Partition.Default` is for commercial regions. Our true default is set as node18 in the `regionalMap` API.

File: packages/@aws-cdk/custom-resource-handlers/test/custom-resources-framework/expected/node-runtime/custom-resource-provider-core.ts
Comment (colifran) [reply packages/@aws-cdk/custom-resource-handlers/test/custom-resources-framework/expected/node-runtime/custom-resource-provider-core.ts:1]: Yes, these are generated by typewriter..

File: packages/@aws-cdk/custom-resource-handlers/lib/custom-resources-framework/classes.ts
Comment (colifran) [reply packages/@aws-cdk/custom-resource-handlers/lib/custom-resources-framework/classes.ts:371]: Agreed. Will move these into constants. Probably as part of the defined modules.

File: packages/@aws-cdk/custom-resource-handlers/lib/custom-resources-framework/classes.ts
Comment (colifran) [reply packages/@aws-cdk/custom-resource-handlers/lib/custom-resources-framework/classes.ts:104]: Correct. The import path changes based on where the file will be compiled. Since alpha modules have a dependency on aws-cdk-lib we need to use that as the import path.

File: packages/@aws-cdk/custom-resource-handlers/lib/custom-resources-framework/callable-expr.ts
Comment (colifran) [reply packages/@aws-cdk/custom-resource-handlers/lib/custom-resources-framework/callable-expr.ts:1]: Extraneous because they are `devDependencies` and eslint will be mad at us.

File: packages/@aws-cdk/aws-amplify-alpha/test/integ.app-asset-deployment.js.snapshot/cdkamplifyappassetdeploymentcomamazonawscdkcustomresourcesamplifyassetdeploymentprovider02396C99.nested.template.json
Comment (colifran) [reply packages/@aws-cdk/aws-amplify-alpha/test/integ.app-asset-deployment.js.snapshot/cdkamplifyappassetdeploymentcomamazonawscdkcustomresourcesamplifyassetdeploymentprovider02396C99.nested.template.json:5]: Correct. If the region is known then we do the lookup and hardcode the runtime. Otherwise, we defer to deploy time with this mapping.

File: packages/@aws-cdk/custom-resource-handlers/test/custom-resources-framework/expected/python-runtime/custom-resource-provider-core.ts
Comment (colifran) [reply packages/@aws-cdk/custom-resource-handlers/test/custom-resources-framework/expected/python-runtime/custom-resource-provider-core.ts:29]: Yeah I also was confused but this is just GitHub showing the diff weird.

File: packages/aws-cdk-lib/core/lib/custom-resource-provider/custom-resource-provider.ts
Comment (colifran) [reply packages/aws-cdk-lib/core/lib/custom-resource-provider/custom-resource-provider.ts:141]: They're marked as deprecated on the `CustomResourceProviderRuntime` enum.

File: packages/@aws-cdk/custom-resource-handlers/lib/custom-resources-framework/modules.ts
Comment (colifran) [reply packages/@aws-cdk/custom-resource-handlers/lib/custom-resources-framework/modules.ts:62]: Agree with the ones used in the `classes.ts`, but these are just used to initialize the `fqn` property in the super class for `ExternalModules`. This would then be used as `this.fqn` for the specific module. That said, I think paths used to initialize the super are okay. I moved everything else into a constant.

File: packages/aws-cdk-lib/aws-lambda/test/function.test.ts
Comment (colifran) [reply packages/aws-cdk-lib/aws-lambda/test/function.test.ts]: Yes. Good suggestions. Updated with new unit tests. We now have unit tests for commercial regions, adc regions, china regions, govcloud regions, and unsupported regions to test that the default runtime is used.

File: packages/aws-cdk-lib/core/test/custom-resource-provider/custom-resource-provider.test.ts
Comment (colifran) [reply packages/aws-cdk-lib/core/test/custom-resource-provider/custom-resource-provider.test.ts:627]: Yeah. Didn't fully understand the difference in region formatting. Updated.

File: packages/aws-cdk-lib/region-info/build-tools/fact-tables.ts
Comment (colifran) [reply packages/aws-cdk-lib/region-info/build-tools/fact-tables.ts:3458]: `Partition.Default` refers to the default partition not the default runtime.

File: packages/@aws-cdk/custom-resource-handlers/test/custom-resources-framework/expected/python-runtime/custom-resource-provider-core.ts
Comment (colifran) [reply packages/@aws-cdk/custom-resource-handlers/test/custom-resources-framework/expected/python-runtime/custom-resource-provider-core.ts:29]: I created new tests specific to python runtimes and node runtimes. I'm not sure why the diff is showing up like this.

File: packages/@aws-cdk/aws-amplify-alpha/scripts/airlift-custom-resource-handlers.sh
Comment (colifran) [reply packages/@aws-cdk/aws-amplify-alpha/scripts/airlift-custom-resource-handlers.sh:14]: Added docstrings and a comment at the top of the file explaining what it does

File: packages/@aws-cdk/custom-resource-handlers/lib/custom-resources-framework/callable-expr.ts
Comment (colifran) [reply packages/@aws-cdk/custom-resource-handlers/lib/custom-resources-framework/callable-expr.ts:8]: Additional unit tests added.

File: packages/@aws-cdk/aws-amplify-alpha/test/integ.app-asset-deployment.js.snapshot/cdkamplifyappassetdeploymentcomamazonawscdkcustomresourcesamplifyassetdeploymentprovider02396C99.nested.template.json
Comment (TheRealAmazonKendra) [reply packages/@aws-cdk/aws-amplify-alpha/test/integ.app-asset-deployment.js.snapshot/cdkamplifyappassetdeploymentcomamazonawscdkcustomresourcesamplifyassetdeploymentprovider02396C99.nested.template.json:5]: Excellent

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
What's still missing here is a default in the map so that if someone attempts to deploy to a region we have not yet added and they are not hardcoding that region in their stack configuration, they are not blocked by us.

**Comment by colifran** [comment]
@TheRealAmazonKendra Looks like CloudFormation support for default values for `FindInMap` is supported by us. I've updated the code to support the default in the deploy time mapping if a region isn't detected.

**Review by TheRealAmazonKendra — APPROVED** [review]

**Pull request merged**
